### PR TITLE
Spark: Add support for describing/showing views

### PR DESCRIPTION
**Pull request by nastra**
This adds support for:
* `DESCRIBE <viewName>` / `DESCRIBE EXTENDED <viewName>`
* `SHOW VIEWS` / `SHOW VIEWS LIKE <pattern>`
* `SHOW TBLPROPERTIES <viewName>`
* `SHOW CREATE TABLE <viewName>`

### CODE REVIEW

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala [hunk 1]
```diff
@@ -123,6 +132,47 @@ case class ExtendedDataSourceV2Strategy(spark: SparkSession) extends Strategy wi
...
+        allowExisting = false,
+        replace = true) :: Nil
+
+    case SetViewProperties(ResolvedV2View(catalog, ident, _), properties) =>
+      val changes = properties.map {
+        case (property, value) => ViewChange.setProperty(property, value)
+      }.toSeq
+      AlterV2ViewExec(catalog, ident, changes) :: Nil
+
+    case UnsetViewProperties(ResolvedV2View(catalog, ident, view), propertyKeys, ifExists) =>
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:155]: Can you remove the unrelated changes?

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateV2ViewExec.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    showColumns(view, builder)
+    showComment(view, builder)
+    showProperties(view, builder)
+    builder ++= s"AS\n${view.query}\n"
+
+    Seq(toCatalystRow(builder.toString))
+  }
+
+  private def showColumns(view: View, builder: StringBuilder): Unit = {
+    val columns = view.schema().fields
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateV2ViewExec.scala:45]: This should only be produced if it was part of the original `CREATE VIEW` command. I think we probably need to store these somewhere in `ViewVersion`.

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteViewCommands.scala [hunk 3]
```diff
@@ -60,17 +65,27 @@ case class RewriteViewCommands(spark: SparkSession) extends Rule[LogicalPlan] wi
...
+      a.copy(query = q)
+
+    case u@UnresolvedView(IsIcebergView(resolved), _, _, _) =>
+      loadView(resolved.catalog, resolved.identifier)
+        .map(v => ResolvedV2View(resolved.catalog.asViewCatalog, resolved.identifier, v))
+        .getOrElse(u)
+
+    case ShowViews(UnresolvedNamespace(ns), pattern, output)
+      if isViewCatalog(spark.sessionState.catalogManager.currentCatalog) =>
+      ShowIcebergViews(UnresolvedNamespace(ns), pattern, output)
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteViewCommands.scala:81]: Shouldn't this resolve the namespace first? This could be called for cases like `SHOW VIEWS IN prod.db`, where the `UnresolvedNamespace` has both `prod` (a catalog) and `db` (a namespace). This needs to run `CatalogAndNamespace.unapply` to get the correct catalog and namespace.

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala (hunk 1 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala]: done

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteViewCommands.scala (hunk 3 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteViewCommands.scala]: I've updated this to use `CatalogAndNamespace` for this particular case. For the case where no namespace is given, we need to match `ShowViews(UnresolvedNamespace(Seq()), pattern, output)`

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateV2ViewExec.scala (hunk 2 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateV2ViewExec.scala:45]: I ran into this case when I was implementing this so I compared this with how `SHOW CREATE TABLE <tbl>` does it. [Here](https://github.com/apache/spark/blob/9088249f02543c549e0b330812c77edcdbbeb795/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateTableExec.scala#L64-L67) it can be seen that it just takes the latest schema of the table and uses it for the output. So should we be aligning `SHOW CREATE VIEW` with `SHOW CREATE TABLE` here?

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteViewCommands.scala [hunk 4]
```diff
@@ -60,6 +63,13 @@ case class RewriteViewCommands(spark: SparkSession) extends Rule[LogicalPlan] wi
         properties = properties,
         allowExisting = allowExisting,
         replace = replace)
+
+    case ShowViews(UnresolvedNamespace(Seq()), pattern, output) if isViewCatalog(catalogManager.currentCatalog) =>
+      ShowIcebergViews(ResolvedNamespace(catalogManager.currentCatalog, Seq.empty), pattern, output)
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteViewCommands.scala:68]: Won't this case be handled the same way by running `CatalogAndNamespace.unapply(Seq())` in the next rule?

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala [hunk 5]
```diff
@@ -123,6 +128,18 @@ case class ExtendedDataSourceV2Strategy(spark: SparkSession) extends Strategy wi
         allowExisting = allowExisting,
         replace = replace) :: Nil
 
+    case d@DescribeRelation(ResolvedV2View(catalog, ident), _, isExtended, _) =>
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:131]: I understand not wanting to bother with `output`, but I think the best practice is to preserve `output` from the logical plan to the physical plan. That's because resolution has occurred with the IDs from the logical plan and we don't want to reassign internal IDs in the physical plan. They should always match. That's why most Spark plans (like `DescribeTableExec`) copy and reuse the output attrs.

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala [hunk 6]
```diff
@@ -123,6 +128,18 @@ case class ExtendedDataSourceV2Strategy(spark: SparkSession) extends Strategy wi
...
+      DescribeV2ViewExec(d.output, catalog.loadView(ident), isExtended) :: Nil
+
+    case show@ShowTableProperties(ResolvedV2View(catalog, ident), propertyKey, _) =>
+      ShowV2ViewPropertiesExec(show.output, catalog.loadView(ident), propertyKey) :: Nil
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:135]: It's a little odd to use `show.output` but then name `propertyKey` in the pattern. Why not name `output` in the pattern and reference it directly?

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java [hunk 7]
```diff
@@ -1149,6 +1149,159 @@ public void createViewWithSubqueryExpressionUsingGlobalTempView() {
                 "because it references to the temporary object global_temp.%s", globalTempView));
   }
 
+  @Test
+  public void describeView() {
+    String viewName = "describeView";
+
+    sql("CREATE VIEW %s AS SELECT id, data FROM %s WHERE id <= 3", viewName, tableName);
+    assertThat(sql("DESCRIBE %s", viewName))
+        .containsExactlyInAnyOrder(row("id", "int", ""), row("data", "string", ""));
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1158]: I think `InAnyOrder` is incorrect. Don't we want these to be in the schema order?

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DescribeV2ViewExec.scala [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    val viewCatalogAndNamespace: Seq[String] = view.currentCatalog +: view.currentNamespace.toSeq
+    val viewProperties = properties.toSeq.sortBy(_._1).map {
+      case (key, value) =>
+        s"'${escapeSingleQuotedString(key)}' = '${escapeSingleQuotedString(value)}'"
+    }.mkString("[", ", ", "]")
+
+
+    toCatalystRow("# Detailed View Information", "", "") ::
+      toCatalystRow("Comment", view.properties.getOrDefault(ViewCatalog.PROP_COMMENT, ""), "") ::
+      toCatalystRow("View Text", view.query, "") ::
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DescribeV2ViewExec.scala:67]: Do we want to include the view text? I think this should be suppressed in favor of showing it for `SHOW CREATE`. Otherwise the output will be formatted really strangely.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java [hunk 9]
```diff
@@ -1149,6 +1149,159 @@ public void createViewWithSubqueryExpressionUsingGlobalTempView() {
...
+            row(NAMESPACE.toString(), "prefixV3", false),
+            row(NAMESPACE.toString(), "v1", false),
+            tempView);
+
+    assertThat(sql("SHOW VIEWS IN %s", catalogName))
+        .contains(
+            row(NAMESPACE.toString(), "prefixV2", false),
+            row(NAMESPACE.toString(), "prefixV3", false),
+            row(NAMESPACE.toString(), "v1", false),
+            tempView);
```
[the remark this answers is not in the document]
Comment (rdblue) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1239]: Also weird when filtering by NAMESPACE below.

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java [hunk 10]
```diff
@@ -1149,6 +1149,159 @@ public void createViewWithSubqueryExpressionUsingGlobalTempView() {
...
+
+    assertThat(sql("SHOW VIEWS IN %s.%s", catalogName, "default"))
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1241]: Instead of "default", should this use `NAMESPACE.toString()`?

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java [hunk 11]
```diff
@@ -1149,6 +1149,159 @@ public void createViewWithSubqueryExpressionUsingGlobalTempView() {
...
+  @Test
+  public void showCreateSimpleView() {
+    String viewName = "showCreateSimpleView";
+    String sql = String.format("SELECT id, data FROM %s WHERE id <= 3", tableName);
+
+    sql("CREATE VIEW %s AS %s", viewName, sql);
+
+    String expected =
+        String.format(
+            "CREATE VIEW %s.%s.%s (id, data) TBLPROPERTIES (\n"
```
Comment (rdblue) [inline spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1272]: Nit: I'd add a newline before `TBLPROPERTIES`.

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateV2ViewExec.scala (hunk 2 above)
Comment (rdblue) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ShowCreateV2ViewExec.scala:45]: I think this is fine for now, but we should consider how to handle the aliases in the future

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
I pointed out a few things that I think I would change, but there is nothing blocking this. +1 overall. Please fix what you agree with (including the `SHOW VIEWS` behavior with temp views) and merge when you're ready!

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala (hunk 6 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ExtendedDataSourceV2Strategy.scala:135]: you're right, this should be the `output` in the pattern for all of these

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java (hunk 7 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1158]: I agree, this should be `containsExactly()`

File: spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nastra) [reply spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1239]: Yes this is the expected Spark behavior as can be seen in https://spark.apache.org/docs/3.5.0/sql-ref-syntax-aux-show-views.html. Spark always lists temp views (which don't belong to any catalog)

File: spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DescribeV2ViewExec.scala (hunk 8 above)
Comment (nastra) [reply spark/v3.5/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DescribeV2ViewExec.scala:67]: I think you're right and it's better to just have `SHOW CREATE` display the query. I've removed the view text from here

**Comment by nastra** [comment]
thank for reviewing @rdblue, I addressed everything and will go ahead and merge this
